operator<< broken for xgenerator

As pointed out by @edgarriba, the following code does not compile:

```cpp
auto xrange = xt::arange(0, 2);
auto yrange = xt::arange(0, 2);

auto grid = xt::meshgrid(xrange, yrange);
std::cout << std::get<0>(grid) << std::endl;
```

On VS2015, for the 0.10.6 version, the error message from the compiler is:

```
1>C:\Users\jmabille\dev\xtensor\include\xtensor/xbuilder.hpp(415): error C2100: illegal indirection
1>  c:\users\jmabille\dev\xtensor\include\xtensor\xgenerator.hpp(193): note: see reference to function template instantiation 'int xt::detail::repeat_impl<xt::xgenerator<Functor,int,shape_type> &>::operator ()<>(void) const' being compiled
1>          with
1>          [
1>              Functor=xt::detail::arange_impl<int>
1>          ]
1>  c:\users\jmabille\dev\xtensor\include\xtensor\xgenerator.hpp(193): note: see reference to function template instantiation 'int xt::detail::repeat_impl<xt::xgenerator<Functor,int,shape_type> &>::operator ()<>(void) const' being compiled
1>          with
1>          [
1>              Functor=xt::detail::arange_impl<int>
1>          ]
1>  C:\Users\jmabille\dev\xtensor\include\xtensor/xio.hpp(197): note: see reference to function template instantiation 'int xt::xgenerator<Functor,int,std::array<std::size_t,2>>::operator ()<>(void) const' being compiled
1>          with
1>          [
1>              Functor=xt::detail::repeat_impl<xt::xgenerator<xt::detail::arange_impl<int>,int,shape_type> &>
1>          ]
1>  C:\Users\jmabille\dev\xtensor\include\xtensor/xio.hpp(197): note: see reference to function template instantiation 'int xt::xgenerator<Functor,int,std::array<std::size_t,2>>::operator ()<>(void) const' being compiled
1>          with
1>          [
1>              Functor=xt::detail::repeat_impl<xt::xgenerator<xt::detail::arange_impl<int>,int,shape_type> &>
1>          ]
1>  C:\Users\jmabille\dev\xtensor\include\xtensor/xio.hpp(657): note: see reference to function template instantiation 'void xt::detail::recurser<2>::run<xt::detail::printer<E,void>,E>(F &,const E &,std::size_t)' being compiled
1>          with
1>          [
1>              E=xt::xgenerator<xt::detail::repeat_impl<xt::xgenerator<xt::detail::arange_impl<int>,int,shape_type> &>,int,std::array<std::size_t,2>>,
1>              F=xt::detail::printer<xt::xgenerator<xt::detail::repeat_impl<xt::xgenerator<xt::detail::arange_impl<int>,int,shape_type> &>,int,std::array<std::size_t,2>>,void>
1>          ]
1>  C:\Users\jmabille\dev\xtensor\include\xtensor/xio.hpp(657): note: see reference to function template instantiation 'void xt::detail::recurser<2>::run<xt::detail::printer<E,void>,E>(F &,const E &,std::size_t)' being compiled
1>          with
1>          [
1>              E=xt::xgenerator<xt::detail::repeat_impl<xt::xgenerator<xt::detail::arange_impl<int>,int,shape_type> &>,int,std::array<std::size_t,2>>,
1>              F=xt::detail::printer<xt::xgenerator<xt::detail::repeat_impl<xt::xgenerator<xt::detail::arange_impl<int>,int,shape_type> &>,int,std::array<std::size_t,2>>,void>
1>          ]
1>  C:\Users\jmabille\dev\xtensor\include\xtensor/xio.hpp(669): note: see reference to function template instantiation 'std::ostream &xt::pretty_print<xt::xgenerator<Functor,int,std::array<std::size_t,2>>>(const xt::xexpression<xt::xgenerator<Functor,int,std::array<std::size_t,2>>> &,std::ostream &)' being compiled
1>          with
1>          [
1>              Functor=xt::detail::repeat_impl<xt::xgenerator<xt::detail::arange_impl<int>,int,shape_type> &>
1>          ]
1>  ..\..\xtensor\test\test_xbuilder.cpp(423): note: see reference to function template instantiation 'std::ostream &xt::operator <<<xt::xgenerator<Functor,int,std::array<std::size_t,2>>>(std::ostream &,const xt::xexpression<xt::xgenerator<Functor,int,std::array<std::size_t,2>>> &)' being compiled
1>          with
1>          [
1>              Functor=xt::detail::repeat_impl<xt::xgenerator<xt::detail::arange_impl<int>,int,shape_type> &>
1>          ]
```

--- a/README.md
+++ b/README.md
@@ -20,7 +20,7 @@ Multi-dimensional arrays with broadcasting and lazy computing.
 
 Containers of `xtensor` are inspired by [NumPy](http://www.numpy.org), the Python array programming library. **Adaptors** for existing data structures to be plugged into our expression system can easily be written.
 
-In fact, `xtensor` can be used to **process `numpy` data structures inplace** using Python's [buffer protocol](https://docs.python.org/3/c-api/buffer.html). Similarly, we can operate on Julia arrays. For more details on the numpy bindings and Julia, check out the [xtensor-python](https://github.com/QuantStack/xtensor-python) and [xtensor-julia](https://github.com/QuantStacl/Xtensor.jl) projects respectively.
+In fact, `xtensor` can be used to **process `numpy` data structures inplace** using Python's [buffer protocol](https://docs.python.org/3/c-api/buffer.html). Similarly, we can operate on Julia and R arrays. For more details on the numpy, Julia and R bindings, check out the [xtensor-python](https://github.com/QuantStack/xtensor-python), [xtensor-julia](https://github.com/QuantStack/Xtensor.jl) and [xtensor-r](https://github.com/QuantStack/xtensor-r) projects respectively.
 
 `xtensor` requires a modern C++ compiler supporting C++14. The following C++ compilers are supported:
 

--- a/include/xtensor/xbroadcast.hpp
+++ b/include/xtensor/xbroadcast.hpp
@@ -274,7 +274,7 @@ namespace xt
      * Returns a constant reference to the element at the specified position in the expression.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater
+     * The number of indices in the sequence should be equal to or greater
      * than the number of dimensions of the function.
      */
     template <class CT, class X>

--- a/include/xtensor/xbuilder.hpp
+++ b/include/xtensor/xbuilder.hpp
@@ -412,7 +412,7 @@ namespace xt
             template <class... Args>
             value_type operator()(Args... args) const
             {
-                std::array<size_type, sizeof...(Args)> args_arr({static_cast<size_type>(args)...});
+                std::array<size_type, sizeof...(Args)> args_arr = {static_cast<size_type>(args)...};
                 return m_source(args_arr[m_axis]);
             }
 

--- a/include/xtensor/xcontainer.hpp
+++ b/include/xtensor/xcontainer.hpp
@@ -428,7 +428,7 @@ namespace xt
      * Returns a reference to the element at the specified position in the container.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater
+     * The number of indices in the sequence should be equal to or greater
      * than the number of dimensions of the container.
      */
     template <class D>
@@ -443,7 +443,7 @@ namespace xt
      * Returns a reference to the element at the specified position in the container.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater
+     * The number of indices in the sequence should be equal to or greater
      * than the number of dimensions of the container.
      */
     template <class D>

--- a/include/xtensor/xexception.hpp
+++ b/include/xtensor/xexception.hpp
@@ -104,13 +104,13 @@ namespace xt
 
     namespace detail
     {
-        template <class S, size_t dim>
+        template <class S, std::size_t dim>
         inline void check_index_impl(const S&)
         {
         }
 
-        template <class S, size_t dim, class... Args>
-        inline void check_index_impl(const S& shape, size_t i, Args... args)
+        template <class S, std::size_t dim, class... Args>
+        inline void check_index_impl(const S& shape, std::size_t arg, Args... args)
         {
             if (sizeof...(Args) + 1 > shape.size())
             {
@@ -118,9 +118,9 @@ namespace xt
             }
             else
             {
-                if (i >= shape[dim] && shape[dim] != 1)
+                if (arg >= shape[dim] && shape[dim] != 1)
                 {
-                    throw std::out_of_range("index " + std::to_string(i) + " is out of bounds for axis "
+                    throw std::out_of_range("index " + std::to_string(arg) + " is out of bounds for axis "
                         + std::to_string(dim) + " with size " + std::to_string(shape[dim]));
                 }
                 check_index_impl<S, dim + 1>(shape, args...);
@@ -139,7 +139,7 @@ namespace xt
     {
         auto dst = static_cast<typename S::size_type>(last - first);
         It efirst = last - std::min(shape.size(), dst);
-        size_t axis = 0;
+        std::size_t axis = 0;
         while (efirst != last)
         {
             if (*efirst >= shape[axis] && shape[axis] != 1)

--- a/include/xtensor/xfunction.hpp
+++ b/include/xtensor/xfunction.hpp
@@ -454,7 +454,7 @@ namespace xt
      * Returns a constant reference to the element at the specified position in the function.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater
+     * The number of indices in the sequence should be equal to or greater
      * than the number of dimensions of the container.
      */
     template <class F, class R, class... CT>

--- a/include/xtensor/xfunctorview.hpp
+++ b/include/xtensor/xfunctorview.hpp
@@ -530,7 +530,7 @@ namespace xt
      * Returns a reference to the element at the specified position in the expression.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater
+     * The number of indices in the sequence should be equal to or greater
      * than the number of dimensions of the function.
      */
     template <class F, class CT>
@@ -575,7 +575,7 @@ namespace xt
      * Returns a constant reference to the element at the specified position in the expression.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater
+     * The number of indices in the sequence should be equal to or greater
      * than the number of dimensions of the function.
      */
     template <class F, class CT>

--- a/include/xtensor/xgenerator.hpp
+++ b/include/xtensor/xgenerator.hpp
@@ -209,7 +209,7 @@ namespace xt
      * Returns a constant reference to the element at the specified position in the function.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater
+     * The number of indices in the sequence should be equal to or greater
      * than the number of dimensions of the container.
      */
     template <class F, class R, class S>

--- a/include/xtensor/xindexview.hpp
+++ b/include/xtensor/xindexview.hpp
@@ -365,7 +365,7 @@ namespace xt
     /**
      * Returns a reference to the element at the specified position in the xindexview.
      * @param first iterator starting the sequence of indices
-     * The number of indices in the squence should be equal to or greater 1.
+     * The number of indices in the sequence should be equal to or greater 1.
      */
     template <class CT, class I>
     template <class It>

--- a/include/xtensor/xreducer.hpp
+++ b/include/xtensor/xreducer.hpp
@@ -423,7 +423,7 @@ namespace xt
      * Returns a constant reference to the element at the specified position in the reducer.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater
+     * The number of indices in the sequence should be equal to or greater
      * than the number of dimensions of the reducer.
      */
     template <class F, class CT, class X>

--- a/include/xtensor/xstridedview.hpp
+++ b/include/xtensor/xstridedview.hpp
@@ -396,7 +396,7 @@ namespace xt
      * Returns a reference to the element at the specified position in the xstrided_view.
      * @param first iterator starting the sequence of indices
      * @param last iterator ending the sequence of indices
-     * The number of indices in the squence should be equal to or greater than the the number
+     * The number of indices in the sequence should be equal to or greater than the the number
      * of dimensions of the container..
      */
     template <class CT, class S, class CD>

--- a/include/xtensor/xstrides.hpp
+++ b/include/xtensor/xstrides.hpp
@@ -30,7 +30,7 @@ namespace xt
     size_type data_offset(const S& strides) noexcept;
 
     template <class size_type, class S, size_t dim = 0, class... Args>
-    size_type data_offset(const S& strides, size_type i, Args... args) noexcept;
+    size_type data_offset(const S& strides, size_type arg, Args... args) noexcept;
 
     template <class size_type, class S, class It>
     size_type element_offset(const S& strides, It first, It last) noexcept;
@@ -71,7 +71,7 @@ namespace xt
     inline auto compute_size(const shape_type& shape) noexcept
     {
         using size_type = std::decay_t<typename shape_type::value_type>;
-        return std::accumulate(shape.begin(), shape.end(), size_type(1), std::multiplies<size_type>());
+        return std::accumulate(shape.cbegin(), shape.cend(), size_type(1), std::multiplies<size_type>());
     }
 
     template <class size_type, class S, size_t dim>
@@ -80,13 +80,13 @@ namespace xt
         return 0;
     }
 
-    template <class size_type, class S, size_t dim, class... Args>
-    inline size_type data_offset(const S& strides, size_type i, Args... args) noexcept
+    template <class size_type, class S, std::size_t dim, class... Args>
+    inline size_type data_offset(const S& strides, size_type arg, Args... args) noexcept
     {
         if (sizeof...(Args) + 1 > strides.size())
             return data_offset<size_type, S, dim>(strides, args...);
         else
-            return i * strides[dim] + data_offset<size_type, S, dim + 1>(strides, args...);
+            return arg * strides[dim] + data_offset<size_type, S, dim + 1>(strides, args...);
     }
 
     template <class size_type, class S, class It>
@@ -94,7 +94,7 @@ namespace xt
     {
         auto dst = static_cast<typename S::size_type>(std::distance(first, last));
         It efirst = last - std::min(strides.size(), dst);
-        return std::inner_product(efirst, last, strides.begin(), size_type(0));
+        return std::inner_product(efirst, last, strides.cbegin(), size_type(0));
     }
 
     namespace detail

--- a/test/test_xbuilder.cpp
+++ b/test/test_xbuilder.cpp
@@ -11,7 +11,7 @@
 #include "xtensor/xarray.hpp"
 
 #include "xtensor/xio.hpp"
-#include <iostream>
+#include <sstream>
 
 namespace xt
 {
@@ -238,6 +238,15 @@ namespace xt
         xarray<double> expect1 = {{0, 1}, {0, 1}, {0, 1}};
         ASSERT_TRUE(all(equal(std::get<0>(mesh), expect0)));
         ASSERT_TRUE(all(equal(std::get<1>(mesh), expect1)));
+    }
+
+    TEST(xbuilder, meshgrid_arange)
+    {
+        auto xrange = xt::arange(0, 2);
+        auto yrange = xt::arange(0, 2);
+        auto grid = xt::meshgrid(xrange, yrange);
+        std::ostringstream stream;
+        stream << std::get<0>(grid) << std::endl;
     }
 
     TEST(xbuilder, triu)

